### PR TITLE
feat: support gradient accumulation in spark torch estimator

### PR DESCRIPTION
**Pull request by thinkall**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

We can use `backward_passes_per_step` in [torch](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_imagenet_resnet50.py#L273-L278), although we need to adjust the [training code](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_imagenet_resnet50.py#L73-L86) to make it work as expected. However, in [spark torch estimator](https://github.com/horovod/horovod/blob/master/horovod/spark/torch/estimator.py), there is no implementation for supporting it.

In this PR, we added support to spark torch estimator, thus we can apply gradient accumlation by simply set up param `backward_passes_per_step` in TorchEstimator. [Example](https://github.com/horovod/horovod/blob/master/examples/spark/pytorch/pytorch_spark_mnist.py) and [test](https://github.com/horovod/horovod/blob/master/test/integration/test_spark_torch.py) are modified accordingly.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by EnricoMi** [comment]
Please rebase with latest master, which includes a fix for the CI.

**Comment by thinkall** [comment]
> Please rebase with latest master, which includes a fix for the CI.

Thanks @EnricoMi , done!

**Comment by chongxiaoc** [comment]
OOO this week. I will take a look when I'm back.

File: horovod/spark/torch/remote.py [hunk 1]
```diff
@@ -194,6 +195,7 @@ def train(serialized_model, optimizer_cls, model_opt_state_serialized,
             # Pass the compression arg only if it is specified by the user.
             dist_optimizer_args['compression'] = gradient_compression
         # Horovod: wrap optimizer with DistributedOptimizer.
+        dist_optimizer_args['backward_passes_per_step'] = backward_passes_per_step
```
Comment (Tixxx) [inline horovod/spark/torch/remote.py:198]: I just realized both torch distributed optimizer and this estimator don't check the value for this. Could you add checks in those places to make sure they are always > 0? Since you are doing division with this number, we need to careful.

File: horovod/spark/torch/remote.py [hunk 2]
```diff
@@ -454,13 +458,16 @@ def _validate(epoch):
...
-        optimizer.zero_grad()
+    def train_minibatch(model, optimizer, transform_outputs, loss_fn, inputs, labels, sample_weights, backward_passes_per_step, batch_idx):
+        if batch_idx % backward_passes_per_step == 0:
+            if batch_idx != 0:
+                optimizer.step()
+            optimizer.zero_grad()
         outputs = model(*inputs)
         outputs, labels = transform_outputs(outputs, labels)
         loss = loss_fn(outputs, labels, sample_weights)
+        loss.div_(float(backward_passes_per_step))
```
Comment (Tixxx) [inline horovod/spark/torch/remote.py:469]: Would be great if we do this only for accumulation >1, we dont want to add new computation nodes for default behavior.

Comment (Tixxx) [reply horovod/spark/torch/remote.py:469]: sometimes we will convert gradients to fp16 for allreduce, this pre-division might have an impact on the numerical precision. I suggest to do this after accumulation i.e before weight update()

Comment (thinkall) [reply horovod/spark/torch/remote.py]: > sometimes we will convert gradients to fp16 for allreduce, this pre-division might have an impact on the numerical precision. I suggest to do this after accumulation i.e before weight update()

If I understand it correctly, allreduce is applied to gradients only, so I guess we need to do loss div before accumulation in order to obtain the correct gradients?

Comment (Tixxx) [reply horovod/spark/torch/remote.py]: ideally (loss / n) translates to (gradients / n), that's why we can linearly scale learning rate directly when adding more workers, but pre-division might cause underflow because of small gradients. This is minor for now as we dont plan to introduce fp16 allreduce for estimators anytime soon.

**Review by Tixxx — APPROVED** [review]

**Review by chongxiaoc — APPROVED** [review]

File: horovod/spark/torch/remote.py (hunk 2 above)
Comment (thinkall) [reply horovod/spark/torch/remote.py]: > ideally (loss / n) translates to (gradients / n), that's why we can linearly scale learning rate directly when adding more workers, but pre-division might cause underflow because of small gradients. This is minor for now as we dont plan to introduce fp16 allreduce for estimators anytime soon.

Thanks for the clarification.
